src/tokens/TokenImport.tsx

```diff
@@ -27,58 +30,175 @@ import variables from 'src/styles/variables'
 
+interface TokenDetails {
+  address: string
+  symbol: string
+  decimals: number
+  name: string
+  balance: bigint
+}
+
+enum Errors {
+  AlreadyImported = 'AlreadyImported',
+  AlreadySupported = 'AlreadySupported',
+  NotContract = 'NotContract',
+  Timeout = 'Timeout',
+  NotERC20 = 'NotERC20',
+}
+class NotContractError extends Error {
+  name = 'NotContractError'
+}
+
 export default function TokenImportScreen(_: Props) {
   const { t } = useTranslation()
   const dispatch = useDispatch()
 
-  const [isCompleteAddress, setIsCompleteAddress] = useState(false)
   const [tokenAddress, setTokenAddress] = useState('')
   const [tokenSymbol, setTokenSymbol] = useState('')
   const [error, setError] = useState<string | null>(null)
   const [networkId] = useState(networkConfig.defaultNetworkId)
+
+  const walletAddress = useSelector(walletAddressSelector)
   const supportedTokens = useSelector((state) => tokensByIdSelector(state, [networkId]))
 
-  const validateAddress = (address: string) => {
-    if (!address) return
-
-    if (isAddress(address)) {
-      setIsCompleteAddress(true)
-      // TODO(RET-891): if already imported, set state as AddressState.AlreadyImported
-      const tokenId = getTokenId(networkId, address.toLowerCase())
-      if (supportedTokens[tokenId]) {
-        setError(t('tokenImport.error.alreadySupported'))
-      } else {
-        setError(null)
-      }
-    } else {
+  const validateAddress = (address: string): address is Address => {
+    if (!address) return false
+
+    if (!isAddress(address)) {
       setError(t('tokenImport.error.invalidToken'))
+      return false
+    }
+
+    // TODO(RET-891): if already imported, set error
+    const tokenId = getTokenId(networkId, address.toLowerCase())
+    if (supportedTokens[tokenId]) {
+      setError(t('tokenImport.error.invalidToken'))
+      ValoraAnalytics.track(AssetsEvents.import_token_error, {
+        networkId,
+        tokenId,
+        tokenAddress,
+        error: Errors.AlreadySupported,
+      })
+      return false
     }
-  }
 
-  const handleAddressFocus = () => {
-    setIsCompleteAddress(false)
     setError(null)
+    return true
   }
 
+  const fetchTokenDetails = async (): Promise<TokenDetails> => {
+    if (!isAddress(tokenAddress)) {
+      // shouldn't happen as this function is called only after validating the address
+      throw new Error('Invalid token address')
+    }
+    if (!walletAddress || !isAddress(walletAddress)) {
+      // should never happen
+      throw new Error('No wallet address found when fetching token details')
+    }
+
+    const client = publicClient[networkIdToNetwork[networkId]]
+    const contract = getContract({
+      abi: erc20.abi,
+      address: tokenAddress,
+      publicClient: client,
+    })
+
+    const contractCode = await client.getBytecode({ address: tokenAddress })
+    if (!contractCode) throw new NotContractError('Contract code is empty')
```

Comment:
this is interesting, but do we need to capture this specific error? if the contract is empty, the fetches for symbol and decimals etc would also fail right? i guess i'm wondering what we would learn from seeing that users are importing tokens that have empty contracts

Reply:
Hmm, perhaps I misunderstood the earlier comment - https://github.com/valora-inc/wallet/pull/4618#discussion_r1422094981 - logging error if not a valid contract - I thought it was about checking if it's a contract and not an ERC-20.

Regarding question: Yes, exactly. We don't really need it. When we check for `symbol`, if that fails, then that's enough for us to check that it implements some of the ERC-20 interface.

I'll remove this bytecode check then and leave only the `symbol()` and the other functions fetching details from ERC-20 contract.